presto-main/src/main/java/com/facebook/presto/operator/SetBuilderOperator.java

```diff
@@ -189,12 +193,30 @@ public boolean isFinished()
         return finished;
     }
 
+    /**
+     * This function processes unfinished work for the current page, and returns if it still
+     * needs input from the previous(child) operator.
+     * @return true if this operator needs additional input pages from its child, false otherwise
+     */
     @Override
     public boolean needsInput()
     {
+        // Since SetBuildOperator is always the last operator (no next operator that needs input from it),
+        // we handle the unfinished work in this method instead of getOutput(). The Driver always
+        // calls it in each iteration.
+        if (unfinishedWork != null) {
+            boolean done = processUnfinishedWork();
+            channelSetBuilder.updateMemoryReservation();
+            if (!done) { return false; }
+        }
         return !finished;
     }
 
+    /**
+     * This function processes one page and add the data to the hash table. If this page can't be fully
+     * consumed (e.g. rehashing fails), the unfinishedWork will be left not null.
+     * @param page The input page from SetBuilderOperator's child(previous) operator
```

Comment:
remove this comment. This is implementation of `Operator#addInput` and if any documentation is missing, it applies there. Implementation details such as "If this page can't be fully consumed (e.g. rehashing fails), the unfinishedWork will be left not null" are better placed in a regular comment.

Reply:
ok